### PR TITLE
Some reformatting of SCE objects before conversion to AnnData

### LINKED ISSUE #368: Add row level metadata to AnnData objects
To make our AnnData output as compliable as possible with CZI, we need to update the existing row/gene metadata present in the AnnData objects to include the necessary entries.
The full documentation can be found here: https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#var-and-rawvar-gene-metadata

In brief we will need to add:

- make sure the index contains Ensembl gene identifiers (I'm pretty sure this is already the case) 
- indication of if features are filtered between raw and normalized 
- Feature biotype ("gene")
- Feature name 
- Feature reference 

Again, we might want to consider if we should include these in the SCE objects prior to converting to AnnData so they are automatically there.

### LINKED ISSUE #417: Add `is_primary_data` to objects before converting to AnnData
We need to include an [`is_primary_data` column](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#is_primary_data) in the cell metadata for AnnData objects. None of our objects should be `is_primary_data` so I think we can just set this to false for all objects prior to converting to AnnData.

### PR DESCRIPTION
**Pull request by allyhawkins**
This PR addresses a few reformatting needs prior to exporting AnnData objects. 

1. Closes #368. I added the one required column for the `rowData` which is the `feature_is_filtered` flag. We don't do any feature filtering between raw and normalized counts matrices, so this is false for everything. Regarding the other `rowData` columns, I think they are to be added by the CZI team and not by us. 
2. In preparation for AlexsLemonade/ScPCA-admin#600, I changed the name of the `participant_id` column to `donor_id`. I chose to do this before adding the sample metadata to the SCE object. This ensures we always refer to the ID as `donor_id` in every object. 
3. Closes #417. I added the `is_primary_data` column to the SCE object with the `FALSE` indication. 
4. One of the last easy things we needed to do was append `X_` to the reduced dim names. We only have those for processed objects, so I added a check for the presence of reduced dimension results. If they are there, I renamed them before converting to AnnData. 

Some references for reviewers: 
- [rowData requirements](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#var-and-rawvar-gene-metadata)
- [colData requirements](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#obs-cell-metadata)
- [Reduced dimension requirements](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#obsm-embeddings)

### CODE REVIEW

**Review by sjspielman — COMMENTED** [review]
The code here all looks fine to me, but I have some questions and overall comments - 

First, regarding overall scope of PR - 
- I'm not sure everything for https://github.com/AlexsLemonade/scpca-nf/issues/368 has been closed here yet. I was expected to see these (untested code, but this is the idea!) additions too:
	- `rowData(sce)$feature_biotype <- "gene"`
	- `rowData(sce)$feature_name <- rownames(sce)`
	- `rowData(sce)$feature_reference <- <something>`
		- At least for now we have all human, so `<something>` would be `"NCBITaxon:9606"`, but we'd want to implement that in a more dynamic way, grabbing the reference organism from the overall metadata and finding the right ID for it
- Just to clarify, this PR is _super not_ intended to address/close #366 
	- I didn't think it was, but since you posted the CZI link just want to confirm!

Next, an organizational comment that might be sort of an annoying lift (sorry...), but in the long-run we will really appreciate having this I think:
It would be good to group in code, as much as possible, the steps we are adding for CZI compliancy. But, that's not always possible for various reasons! So, at a minimum, I'd suggest adding comments throughout the code to indicate when something is _specifically for CZI compliancy_, e.g. `# added for CZI compliancy`. This will be particularly helpful for spots where we modify metadata in ways that _we_ don't need, but CZI wants, like `sce$is_primary_data <- FALSE`.


Finally, a more general comment about CZI compliancy - do we have a single overall checklist or so tracking everything that needs to be done and associated progress? I ask this because, for example, the append `X_` to reducedDims isn't an issue, but clearly PR this accomplishes it! So, for this comment, I'm not asking for anything to change in this PR but to note that project-wise, we'd benefit from, for example, a big comment with a checklist in https://github.com/AlexsLemonade/ScPCA-admin/issues/572 or something like that to make sure everything is recorded. Maybe you have this going on your end already, or it's somewhere else and I missed it though!

**Comment by allyhawkins** [comment]
> * I'm not sure everything for [Add row level metadata to AnnData objects #368](https://github.com/AlexsLemonade/scpca-nf/issues/368) has been closed here yet. I was expected to see these (untested code, but this is the idea!) additions too:
>   
>   * `rowData(sce)$feature_biotype <- "gene"`
>   * `rowData(sce)$feature_name <- rownames(sce)`
>   * `rowData(sce)$feature_reference <- <something>`
>     
>     * At least for now we have all human, so `<something>` would be `"NCBITaxon:9606"`, but we'd want to implement that in a more dynamic way, grabbing the reference organism from the overall metadata and finding the right ID for it

So, when I originally made the issue for adding the rowData modifications, I was mistaken and thought we were supposed to add all of the columns that you mentioned. However, when going back to work on it I noticed that the only additions that are to be added by the `curator` (Data Lab) are making sure the ensembl id's are the rownames and adding the `feature_is_filtered` column. The other additions are to be annotated by the `Data Portal` (CZI). 

Screenshot 2023-08-31 at 9 53 32 AM

> * Just to clarify, this PR is _super not_ intended to address/close [Add Cell level metadata to AnnData objects #366](https://github.com/AlexsLemonade/scpca-nf/issues/366)
>   * I didn't think it was, but since you posted the CZI link just want to confirm!

No, this isn't meant to close it, but it is meant to address the addition of the `donor_id`. Addressing https://github.com/AlexsLemonade/scpca-nf/issues/419 will help address the rest of what isn't found in that issue. 

> Next, an organizational comment that might be sort of an annoying lift (sorry...), but in the long-run we will really appreciate having this I think: It would be good to group in code, as much as possible, the steps we are adding for CZI compliancy. But, that's not always possible for various reasons! So, at a minimum, I'd suggest adding comments throughout the code to indicate when something is _specifically for CZI compliancy_, e.g. `# added for CZI compliancy`. This will be particularly helpful for spots where we modify metadata in ways that _we_ don't need, but CZI wants, like `sce$is_primary_data <- FALSE`.

Thanks for this suggestion! I went ahead and did some reorganization to the code, specifically making a function for making CZI compliant. Within that function all the modifications to the SCE are made to make sure the output object is CZI compliant. That is applied to both the main experiment and any altExp. I did have to reshuffle the order because the altExp doesn't have the `sample_metadata` yet and we want to add that in using the `sample_metadata` in the main experiment. 

> Finally, a more general comment about CZI compliancy - do we have a single overall checklist or so tracking everything that needs to be done and associated progress? I ask this because, for example, the append `X_` to reducedDims isn't an issue, but clearly PR this accomplishes it! So, for this comment, I'm not asking for anything to change in this PR but to note that project-wise, we'd benefit from, for example, a big comment with a checklist in [AlexsLemonade/ScPCA-admin#572](https://github.com/AlexsLemonade/ScPCA-admin/issues/572) or something like that to make sure everything is recorded. Maybe you have this going on your end already, or it's somewhere else and I missed it though!

To hopefully add some clarity to this process, I added a checklist to https://github.com/AlexsLemonade/ScPCA-admin/issues/572 and checked off anything that has been already merged in and is present in the output objects. I also included a link to our slab notes from our design discussion.

File: bin/sce_to_anndata.R [hunk 1]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
   stop("output rna file name must end in .hdf5 or .h5")
 }
 
-# Convert to AnnData -----------------------------------------------------------
+# CZI compliance function ------------------------------------------------------
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:50]: :tada:

File: bin/sce_to_anndata.R [hunk 2]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
 
-# read in sce
-sce <- readr::read_rds(opt$input_sce_file)
+# this function applies any necessary reformatting or changes needed to make
+# sure that the sce that is getting converted to AnnData is compliant with CZI
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:53]: maybe?
```suggestion
# sure that the sce that is getting converted to AnnData is compliant with CZI
# CZI 3.0.0 requirements: https://github.com/chanzuckerberg/single-cell-curation/blob/b641130fe53b8163e50c39af09ee3fcaa14c5ea7/schema/3.0.0/schema.md
```

File: bin/sce_to_anndata.R [hunk 3]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+format_czi <- function(sce){
 
-# grab sample metadata
-sample_metadata <- metadata(sce)$sample_metadata
+  # add library id as a column to the sce object
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:56]:
```suggestion
  # add library_id as an sce colData column
```

File: bin/sce_to_anndata.R [hunk 4]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+  sce$library_id <- metadata(sce)$library_id
 
-# add library id as a column to the sce object
-sce$library_id <- metadata(sce)$library_id
+  # add is_primary_data column only needed for anndata objects
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:59]:
```suggestion
  # add is_primary_data column; only needed for anndata objects
```

File: bin/sce_to_anndata.R [hunk 5]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+  # remove sample metadata from sce metadata, otherwise conflicts with converting object
+  metadata(sce) <- metadata(sce)[names(metadata(sce)) != "sample_metadata"]
+
+  # modify rowData
+  # we don't do any gene filtering between normalized and raw counts matrix
+  # so everything gets set to false
+  rowData(sce)$feature_is_filtered <- FALSE
+
+  # paste X to reduced dim names if present
+  if(!is.null(reducedDimNames(sce))){
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:75]: spacing 
```suggestion
  if (!is.null(reducedDimNames(sce))) {
```

File: bin/sce_to_anndata.R [hunk 6]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+    reducedDimNames(sce) <- glue::glue("X_{reducedDimNames(sce)}")
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:76]: i love `glue::glue()`!

File: bin/sce_to_anndata.R [hunk 7]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+  }
+
+  return(sce)
+
+}
+
+# AltExp to AnnData -----------------------------------------------------------
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:83]: A bit of reorg here to go in order:
- read in SCE
- czi-ify main exp
- czi-ify alt exp

note that subsequent deletion suggestions should bring it all home!

```suggestion

# Read in sce -------------------------------------------
sce <- readr::read_rds(opt$input_sce_file)

# MainExp to AnnData -----------------------------------------------------------

# make main sce czi compliant
sce <- format_czi(sce)

# AltExp to AnnData -----------------------------------------------------------
```

File: bin/sce_to_anndata.R [hunk 8]
```diff
@@ -47,28 +47,43 @@ if(!(stringr::str_ends(opt$output_rna_h5, ".hdf5|.h5"))){
...
+
+# read in sce
+sce <- readr::read_rds(opt$input_sce_file)
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:86]:
```suggestion
```

File: bin/sce_to_anndata.R [hunk 9]
```diff
@@ -106,3 +114,14 @@ if(!is.null(opt$feature_name)){
   )
 
 }
+
+# MainExp to AnnData -----------------------------------------------------------
+
+# make main sce czi compliant
+sce <- format_czi(sce)
```
Comment (sjspielman) [inline bin/sce_to_anndata.R:121]:
```suggestion
```

**Review by sjspielman — COMMENTED** [review]
Ahh thank you for the drawing/arrows, really! I had definitely missed those "tags" when looking at CZI docs, makes total sense!

✅ https://github.com/AlexsLemonade/ScPCA-admin/issues/572 checklist 

I left a couple more small organizational comments, but the code all looks good to me at this point. Caveat, I did _not_ run the code but I'm assuming you are, so let me know if you want me to test it out as well!

One final item mostly for me and/or posterity. I wondered whether it was actually true that `rowData(sce)$feature_is_filtered <- FALSE` for the CITE-seq data. For example, did we remove positive or negative controls at some point and I forgot we did that? I am happy to say that it seems we did not, so this is fine! https://github.com/AlexsLemonade/scpca-nf/blob/bed9effce77e1952bab4577350de336be73da9b2/bin/post_process_sce.R#L180-L230

**Comment by allyhawkins** [comment]
> A bit of reorg here to go in order:
> 
> read in SCE
> czi-ify main exp
> czi-ify alt exp

I had switched it to `altExp` first because we need to use the `metadata(sce)$sample_metadata` from the main SCE. This gets removed when running `format_czi` so it wouldn't be there to use if we did the main one first. But I know it makes more sense for main first, so I just pulled out the `sample_metadata` at the beginning to account for this.

**Review by sjspielman — APPROVED** [review]
🚀

**Pull request merged**